Changes to DynamicStressDivergenceTensors to enable use of rayleigh d…

sveerara opened the pull request:
…amping with StressDivergencePFFracTensors #8351

DynamicStressDivergenceTensors is edited so that it derives from StressDivergencePFFracTensors instead of StressDivergenceTensors. 

combined/tests/phase_field_fracture/crack2d.i is modified to demonstrate the syntax if DynamicTensorMechanicsAction is used to set up the stress divergence kernels. Adding zeta = some real number to this action would turn on stiffness proportional Rayleigh damping.

closes #8351

sveerara committed:
Changes to StressDivergencePFFracTensors to enable use of rayleigh damping with StressDivergencePFFracTensors #8351

sveerara commented:
The tensor_mechanics/capped_weak_plane/pull_and_shear.i and push_and_shear.i tests are failing on civet with a seg fault but all the tests pass on my computer. Could someone please help me figure out why these tests are failing? Thanks!

permcody commented:
@sveerara - if you have access to the HPC Enclave you can login to one of the build boxes to test. If that works for you, I can have @milljm give you direction on how to do that.

sveerara commented:
@permcody - I do have access to the usual HPC systems like falcon and fission. If that access is sufficient, can you please ask @milljm to send me instructions on how to log on to the build boxes and run the tests there? Thanks!

permcody commented on modules/tensor_mechanics/src/kernels/StressDivergencePFFracTensors.C:
```diff
@@ -12,17 +12,24 @@ InputParameters validParams<StressDivergencePFFracTensors>()
+      if (isCoupled("c"))
+      {
+        _c_coupled = true;
+        _c_var = coupled("c");
```
@sveerara - Looks like you deleted the initialization for a few variables. This isn't a good idea in general if you plan to access them. `_c_coupled` for instance can uninitialized and then you test against it in `computeQpOffDiagJacobian`. The other uninitialized variables (`_c_coupled`, `_c_var`) are protected and only evaluated if `_c_coupled` is true but that's still three variables that can be in undefined states.

sveerara replied:
oh okay. I will try initializing it with false values, i.e., _c_coupled = false and _c_var = 0 and then if c is a valid parameter then I can change the values of these variables using this if condition. Thanks!

permcody commented on modules/tensor_mechanics/src/kernels/StressDivergencePFFracTensors.C:
```diff
@@ -32,7 +39,7 @@ StressDivergencePFFracTensors::computeQpOffDiagJacobian(unsigned int jvar)
     Real val = 0.0;
     for (unsigned int k = 0;k < 3; ++k)
-      val += _d_stress_dc[_qp](_component,k) * _grad_test[_i][_qp](k);
+      val += (*_d_stress_dc)[_qp](_component,k) * _grad_test[_i][_qp](k);
```
space after comma

permcody commented on modules/tensor_mechanics/src/kernels/StressDivergencePFFracTensors.C:
```diff
@@ -12,17 +12,24 @@ InputParameters validParams<StressDivergencePFFracTensors>()
-    _c_var(_c_coupled ? coupled("c") : 0),
-    _d_stress_dc(getMaterialPropertyDerivative<RankTwoTensor>(_base_name + "stress", getVar("c", 0)->name()))
+    _c_valid(isParamValid("c")),
+    _d_stress_dc(_c_valid ? &getMaterialPropertyDerivative<RankTwoTensor>(_base_name + "stress", getVar("c", 0)->name()) : NULL)
```
Use `nullptr` instead of `NULL` (C++11)

permcody commented on modules/tensor_mechanics/src/kernels/StressDivergencePFFracTensors.C:
```diff
@@ -12,17 +12,24 @@ InputParameters validParams<StressDivergencePFFracTensors>()
   InputParameters params = validParams<StressDivergenceTensors>();
   params.addClassDescription("Stress divergence kernel for phase-field fracture: Additionally computes off diagonal damage dependent Jacobian components");
-  params.addCoupledVar("c", "Phase field damage variable: Used to indicate calculation of Off Diagonal Jacobian term");
+  params.addCoupledVar("c","Phase field damage variable: Used to indicate calculation of Off Diagonal Jacobian term");
```
space after comma

permcody reviewed (changes requested):
I tried to look for your segfault and indicated a few minor problems you may want to look at.

permcody commented:
Looks like your seg fault went away!

sveerara commented:
Yes, thanks for figuring out the issue @permcody.

bwspenc reviewed (changes requested):
The code changes seem reasonable, and I see how this is helping you do what you want to do (use dynamics together with phase field fracture) but the inheritance here seems a little weird.  You would now be using an object specific to phase field fracture whenever you want to do any type of dynamics analysis.

When we only wanted to use dynamics and phase field fracture independently, I think the inheritance relationship between those objects and the base StressDivergenceTensors object made sense, but it wouldn't anymore.  I'd prefer to solve this problem by composing things together rather than inheriting.

StressDivergencePFFracTensors just adds additional off-diagonal Jacobian entries, and doesn't otherwise change anything about the basic stress divergence.  I think we could just make it no longer inherit from StressDivergenceTensors, and require both of those blocks in the input file.  That would allow you to do what you want to do.

DynamicStressDivergenceTensors is more intertwined with StressDivergenceTensors, so maybe it really does need to inherit from StressDivergenceTensors, but it might be possible to make it an object that you just use in addition to StressDivergenceTensors, and remove that inheritance as well.

Those are just some initial thoughts -- I'm happy to talk through this more.

sveerara commented:
@bwspenc, thanks for your comments on the commit. 

I agree that StressDivergencePFFracTensors does not really have to derive from StressDivergenceTensors because it is editing only the off diagonal Jacobian. I can make it derive it from Kernel. Then both the StressDivergenceTensors and StressDivergencePFFracTensors blocks will be required in the input file for a static analysis and DynamicStressDivergenceTensors and StressDivergencePFFracTensors would be required for a dynamic analysis. 

DynamicStressDivergenceTensors has a lot in common with StressDivergenceTensors. If I make it derive from Kernel instead of StressDivergenceTensors, I would have to copy a lot of the volumetric locking code over. So I think the inheritance in DynamicStressDivergenceTensors from StressDivergenceTensors is required.

bwspenc commented:
@sveerara I think we're in agreement that the immediate solution is to just make StressDivergencePFFracTensors derive from Kernel.

As far as the dynamics kernel goes, does it just add to what is contribued by the standard stress divergence kernel?  If that is the case, you could have both kernels defined in an input file, and they would both add their own parts in.  If it needs to modify that (and maybe it does), then it really should inherit as it does now.

sveerara commented:
@bwspenc The dynamics kernel adds to the contribution from the standard stress divergence kernel. But DynamicStressDivergenceTensors would still have to inherit from StressDivergenceTensors because the Jacobian contributions from the dynamics part is basically the one that is calculated for StressDivergenceTensors (stiffness matrix K) multiplied by a factor.

So I could split into two parts for the input file - one that just calculates Ku and other that calculates the contribution due to HHT time integration and stiffness proportional Rayleigh damping. While splitting the dynamics contributions is good for separating out different parts of the PDE, we would be calling StressDivergenceTensors::computeQpJacobian() twice - once from StressDivergenceTensors and once from DynamicStressDivergenceTensors. So I think it will add more computational effort.

bwspenc commented:
@sveerara That makes sense.  Let's not touch the dynamics kernel.

sveerara commented:
@bwspenc, I have addressed your comments. Thanks!

bwspenc commented:
@sveerara This looks good, and I think this is a much cleaner solution. 

However, as I look at this, now I'm concerned that this will break existing models because the input files will run, but the user wouldn't get any indication that the StressDivergencePFFracTensors is no longer contributing to the displacement residual (except that if they look at their results, they'll see that they have no displacements and their models probably won't even run because there are zeros on the diagonal).

I propose this alternative solution:
Since StressDivergencePFFracTensors will no longer compute the stress divergence, it's really not appropriately named any more.  Let's keep it around unchanged, but deprecate it. 

Then, add a new class that is identical to what you currently have for StressDivergencePFFracTensors, but name it in a way that better indicates that it is for dealing with the coupling between mechanics and phase field.  There's already one that is kind of like that (GeneralizedPlaneStrainOffDiag).  How about PhaseFieldFractureMechanicsOffDiag?

If you can, have the deprecation message for StressDivergencePFFracTensors indicate what the new syntax to be used should be.  I think you can do this by generating a warning in the constructor of that class.

sveerara commented:
@bwspenc, I agree with your comment that users would be confused if StressDivergencePFFracTensors suddenly gives different results without giving an explanation. 

I have left StressDivergencePFFracTensors unchanged with a comment saying that it will be deprecated soon. I have created another class PhaseFieldFractureMechanicsOffDiag which is derived from Kernel and adds the Off diagonal contribution from phase field.

bwspenc commented on modules/tensor_mechanics/src/kernels/StressDivergencePFFracTensors.C:
```diff
@@ -23,6 +23,7 @@ StressDivergencePFFracTensors::StressDivergencePFFracTensors(const InputParamete
     _c_var(_c_coupled ? coupled("c") : 0),
     _d_stress_dc(getMaterialPropertyDerivative<RankTwoTensor>(_base_name + "stress", getVar("c", 0)->name()))
 {
+  mooseDeprecated("StressDivergencePFFracTensors is deprecated. Please use PhaseFieldFractureMechanicsOffDiag instead.");
```
... and StressDivergenceTensors

bwspenc reviewed (changes requested):
One minor other little thing: change the registration of the deprecated object to use `registerDeprecatedObject()`.  There's one example of the usage of a closely related function, `registerDeprecatedObjectName()` in tensor_mechanics.  Give it a date a few months from now for it to expire (and put that on your calendar to remember to remove it!)

sveerara commented:
@bwspenc I have addressed your comments. I have added StressDivergencePFFracTensors as a deprecated object with an associated date and I will remove this class on that date. Thanks!

bwspenc commented on modules/combined/tests/phase_field_fracture/crack2d.i:
```diff
@@ -48,18 +48,20 @@
     G0_var = 'G0_pos'
     dG0_dstrain_var = 'dG0_pos_dstrain'
   [../]
+  [./DynamicTensorMechanics]
```
Did you intend to use DynamicTensorMechanics here instead of TensorMechanics?  I suppose either one works.

sveerara replied:
Well either one works. I had used DynamicTensorMechanics to check if it works with both TensorMechanics and DynamicTensorMechanics. I can change it back to TensorMechanics if needed.

bwspenc replied:
OK, that's fine.  I just wanted to make sure it was intentional.  Someone might pick this up and get confused.

bwspenc commented:
@permcody You requested changes a while back.  This has changed significantly since then.  I'm good with the code as-is.  If you are as well, could you please approve your changes?
